extensions/authn/request_authenticator.cc

```diff
@@ -0,0 +1,122 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "extensions/authn/request_authenticator.h"
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
+#include "common/http/headers.h"
+
+using istio::authn::Payload;
+
+namespace Extensions {
+namespace AuthN {
+
+namespace {
+// The default header name for an exchanged token
+static const std::string kExchangedTokenHeaderName = "ingress-authorization";
+
+// Returns whether the header for an exchanged token is found
+bool FindHeaderOfExchangedToken(
+    const istio::security::v1beta1::JWTRule& jwt_rule) {
+  return (jwt_rule.from_headers_size() == 1 &&
+          Envoy::Http::LowerCaseString(kExchangedTokenHeaderName) ==
+              Envoy::Http::LowerCaseString(jwt_rule.from_headers(0).name()));
+}
+}  // namespace
+
+Envoy::Http::RegisterCustomInlineHeader<
+    Envoy::Http::CustomInlineHeaderRegistry::Type::RequestHeaders>
+    access_control_request_method_handle(
+        Envoy::Http::CustomHeaders::get().AccessControlRequestMethod);
+Envoy::Http::RegisterCustomInlineHeader<
+    Envoy::Http::CustomInlineHeaderRegistry::Type::RequestHeaders>
+    origin_handle(Envoy::Http::CustomHeaders::get().Origin);
+
+bool isCORSPreflightRequest(const Envoy::Http::RequestHeaderMap& headers) {
+  return headers.Method() &&
+         headers.Method()->value().getStringView() ==
+             Envoy::Http::Headers::get().MethodValues.Options &&
+         !headers.getInlineValue(origin_handle.handle()).empty() &&
+         !headers.getInlineValue(access_control_request_method_handle.handle())
+              .empty();
+}
+
+RequestAuthenticator::RequestAuthenticator(
+    FilterContextPtr filter_context,
+    const istio::security::v1beta1::RequestAuthentication& policy)
+    : request_authentication_policy_(policy), filter_context_(filter_context) {}
+
+bool RequestAuthenticator::run(Payload* payload) {
+  if (isCORSPreflightRequest(filter_context_->headerMap())) {
+    // The CORS preflight doesn't include user credentials, allow regardless of
+    // JWT policy. See
+    // http://www.w3.org/TR/cors/#cross-origin-request-with-preflight.
+    return true;
+  }
+
+  absl::string_view path;
+  if (filter_context_->headerMap().Path() != nullptr) {
+    path = filter_context_->headerMap().Path()->value().getStringView();
+
+    // Trim query parameters and/or fragment if present
+    size_t offset = path.find_first_of("?#");
+    if (offset != absl::string_view::npos) {
+      path.remove_suffix(path.length() - offset);
+    }
+  }
+
+  if (payload != nullptr &&
+      payload->payload_case() == Payload::PayloadCase::kJwt) {
+    if (validateJwt(payload->mutable_jwt())) {
+      filter_context_->setOriginResult(payload);
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool RequestAuthenticator::validateJwt(istio::authn::JwtPayload* jwt) {
+  for (const auto& jwt_rule : request_authentication_policy_.jwt_rules()) {
```

Comment:
I feel this could be greatly simplified without depending on the request_authentication policy at all.

Istiod already generates the Envoy JWT filter config from consolidated RequestAuthN policies, Envoy JWT filter should validate and output the JWT token to the metadata (via PayloadInMetadata)

can we just pass a list of issuers in the authn filter's config instead of passing the whole RequestAuthN policy? (I want to avoid dependency on the user-facing API in proxy)